### DISCUSSION TITLE
HyperLogLog compatibility with Apache Spark

### DISCUSSION THREAD

**Discussion by billcrook** [trinodb/trino#24384] [General]
I'm interested in using spark to write HLL data structures that are compatible with Trino (and implicitly aws athena). From [what I can tell](https://github.com/trinodb/trino/blob/97508bb2d182f9a76666ad557c3506948bc0ec80/core/trino-main/src/main/java/io/trino/operator/aggregation/state/HyperLogLogState.java#L16), trino uses HLL datastructures from the associated [airlift](https://github.com/airlift/airlift) project. Spark [leverages](https://github.com/apache/spark/blob/af4f37c94ce47c28e3c324499463e032b68d454d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/datasketchesAggregates.scala#L36) the [apache datasketches](https://datasketches.apache.org/) library for HLL. 

I'm hitting a dead end when trying to find anyone doing this. Before I go down a rabbit hole of writing a custom spark aggregator to use the airlift HLL datastructure, does anyone know of prior art or if there are plans in the trino pipeline to improve this compatibility?

**Comment by hashhar** [comment]
I don't think most people actually store the serialized HLLs into a table.

cc: @electrum if he's thought about interop in the past.

**Reply by electrum** [reply]
This is a question for @martint

**Reply by junwei-wang** [reply]
Actually, materializing HLL is a real usecase when data streaming in and sketches are gradually built for a e.g. hourly/daily time bucket, then at query time, the sketches for time buckets can be merged into big timeframe.

**Comment by billcrook** [comment]
> I don't think most people actually store the serialized HLLs into a table.

My use case is nearly 1T events aggregated down into a reasonable grain which the BI team can then further aggregate across many dimensions. I don't believe it's an unusual use case.

**Reply by convex1** [reply]
+1 this dead end is real. Any solution to this?

**Reply by billcrook** [reply]
@martint Any thoughts on this?

**Comment by martint** [comment]
The only way to support reading HLLs generated by Spark would be to add functions to Trino that can decode them, compute their union and cardinalities. The two implementations are fundamentally different, and there's no way to make them compatible. Minor aspects such as choice of hash function or which bits to use for calculating the bucket result in data structures that cannot be mixed together and produce a sensible result.

**Reply by billcrook** [reply]
Thanks for your input.

The other way to do this would be to create a new spark library to read/write HLLs from Trino. Do you know of any documentation/pointers of how it may be possible to leverage the HLL lib used by Trino?

Also, I'm wondering what the relationship between Trino and Airlift is? Specifically, why was an HLL implementation from a 3rd-party rest services framework chosen and is the plan to keep that dependency going forward?

Again, thanks for you help.

**Reply by martint** [reply]
Some of the original authors of Trino (in particular, Dain, David and I) worked on Airlift before we started Trino and continue to maintain it, so it was natural to use it. I wrote the HLL implementation in Airlift.

**Comment by billcrook** [comment]
It looks like presto has support for [kll sketch functions](https://prestodb.io/docs/current/functions/sketch.html#kll-sketches). If trino had this functionality, I believe it would provide a path for spark-athena HLL compatibility.

**Comment by convex1** [comment]
I ran into this exact problem and solved it about a year ago. I wrote a small **Java library** that **produces Airlift-format HLL sketches directly from Spark** — I run it as a UDF inside our Spark jobs, and the resulting sketches are read and merged natively by Trino and AWS Athena. We have been using it in production for ~1 year to roll up billions events into pre-aggregated sketches that our data engineering teams query downstream, so it's well past the proof-of-concept stage. 
  
Happy to open-source it. My instinct is that a standalone repo / library is the best home — bundling it into the Spark repo would add weight to an already-large project, and it's really a thin compatibility layer rather than a core Spark feature. Let me know what format would be most useful and I am glad to put up a PR or publish the repo.

**Reply by billcrook** [reply]
That would be amazing! Source format would be fine IMO. Kudos for knocking that out
